### PR TITLE
STORM-3590: Adds test to validate that GRAS's node sort is stable and…

### PR DESCRIPTION
**Pull request by govind-menon**
… prevents starvation and fragmentation

### CODE REVIEW

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java [hunk 1]
```diff
@@ -300,7 +302,63 @@ public void testGrasRequiringEviction() {
...
+        Cluster cluster = new Cluster(new INimbusTest(), new ResourceMetrics(new StormMetricsRegistry()), supMap, new HashMap<>(), topologies, conf);
+
+        // should schedule gpu and noGpu successfully
+        scheduler = new ResourceAwareScheduler();
+        nonGpuconf.put(DaemonConfig.RESOURCE_AWARE_SCHEDULER_MAX_TOPOLOGY_SCHEDULING_ATTEMPTS, 1); // allows no evictions
+        scheduler.prepare(nonGpuconf);
+        // schedule gpu topology
+        scheduler.schedule(topologies, cluster);
+
+        cluster.addTopologyToClusterToBeScheduled(createTestStormTopology(stormTopologyNoGpu,  10, noGpu, nonGpuconf));
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java]: It doesn't look like a good idea to add a new function in the regular code solely for unit test. We don't need to add this function to achieve what we want. We can follow this https://github.com/apache/storm/blob/master/storm-server/src/test/java/org/apache/storm/scheduler/resource/TestResourceAwareScheduler.java#L504-L511

File: storm-server/src/main/java/org/apache/storm/scheduler/Topologies.java [hunk 2]
```diff
@@ -76,6 +76,14 @@ public Topologies(Topologies src) {
         return ret;
     }
 
+    public void addTopology(TopologyDetails details) {
```
Comment (kishorvpatil) [inline storm-server/src/main/java/org/apache/storm/scheduler/Topologies.java]: This is supposed to be immutable class. just create another instance of topologies using constructor.

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java [hunk 3]
```diff
@@ -300,7 +302,64 @@ public void testGrasRequiringEviction() {
...
+        Cluster cluster = new Cluster(new INimbusTest(), new ResourceMetrics(new StormMetricsRegistry()), supMap, new HashMap<>(), topologies, conf);
+
+        // should schedule gpu and noGpu successfully
+        scheduler = new ResourceAwareScheduler();
+        nonGpuconf.put(DaemonConfig.RESOURCE_AWARE_SCHEDULER_MAX_TOPOLOGY_SCHEDULING_ATTEMPTS, 1); // allows no evictions
+        scheduler.prepare(nonGpuconf);
+        // schedule gpu topology
+        scheduler.schedule(topologies, cluster);
+
+        topologies.addTopology(createTestStormTopology(stormTopologyNoGpu,  10, noGpu, nonGpuconf));
```
Comment (kishorvpatil) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java]: This is supposed to be immutable class. just create another instance of topologies using constructor.  

try ` Topologies topologies = new Topologies(topo[0], topo[1]);`

File: storm-server/src/main/java/org/apache/storm/scheduler/Topologies.java (hunk 2 above)
Comment (Ethanlm) [reply storm-server/src/main/java/org/apache/storm/scheduler/Topologies.java:79]: We shouldn't add methods to regular code while it's only used in unit test

File: storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java [hunk 4]
```diff
@@ -300,7 +302,64 @@ public void testGrasRequiringEviction() {
...
+        Topologies topologies = new Topologies(createTestStormTopology(stormTopologyWithGpu, 9, gpu, conf));
+
+        Map<String, Double> genericResourcesMap = new HashMap<>();
+        genericResourcesMap.put("gpu.count", 2.0);
+        Map<String, SupervisorDetails> supMap = genSupervisors(4, 4, 200, 2000, genericResourcesMap);
+        Cluster cluster = new Cluster(new INimbusTest(), new ResourceMetrics(new StormMetricsRegistry()), supMap, new HashMap<>(), topologies, conf);
+
+        // should schedule gpu and noGpu successfully
+        scheduler = new ResourceAwareScheduler();
+        nonGpuconf.put(DaemonConfig.RESOURCE_AWARE_SCHEDULER_MAX_TOPOLOGY_SCHEDULING_ATTEMPTS, 1); // allows no evictions
```
Comment (Ethanlm) [inline storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestGenericResourceAwareStrategy.java:348]: This is not able to indicate if eviction happens or not.
Waiting on https://github.com/apache/storm/pull/3213 so we can have a right way to detect eviction.

**Pull request closed without merging**